File: Excel/645 Challenge.py
import pandas as pd
import numpy as np

path = "645 Align WBS Data.xlsx"
input = pd.read_excel(path, usecols="A:B", nrows=30)
test = pd.read_excel(path, usecols="E:J", nrows=30)

def process_data(input):
    for i in range(6):
        col = str(i)
        input[col] = np.where(input['WBS'].str.contains(f'WBS_{col}'), input['WBS'], np.nan)
        input[col] = input[col].ffill() if i == 0 else input.groupby(str(i - 1))[col].ffill()
    return input.fillna("XXX").drop(columns=['WBS', 'ID'])

result = process_data(input)
result.columns = test.columns

print(result.equals(test)) # True

File: Excel/645 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/645 Align WBS Data.xlsx"
input = read_excel(path, range = "A1:B30")
test  = read_excel(path, range = "E1:J30")

result = reduce(0:5, function(data, i) {
  col = as.character(i)
  data = data %>% 
    mutate(!!col := if_else(str_detect(WBS, paste0("WBS_", col)), WBS, NA_character_))
  
  data = if (i == 0) {
    data %>% fill(!!sym(col), .direction = "down")
  } else {
    data %>% group_by(!!sym(as.character(i - 1))) %>% 
      fill(!!sym(col), .direction = "down") %>% 
      ungroup()
  }
  data
}, .init = input) %>% 
  mutate(across(everything(), ~replace_na(.x, "XXX"))) %>% 
  select(-WBS, -ID)

all.equal(result, test, check.attributes = FALSE) # TRUE
